lib/rules/selector-max-specificity/index.cjs

```diff
@@ -27,29 +29,127 @@ const meta = {
 /** @typedef {import('@csstools/selector-specificity').Specificity} Specificity */
 
 /**
- * Return a zero specificity. We need a new instance each time so that it can mutated.
- *
- * @returns {Specificity}
+ * @param {import('postcss-selector-parser').Selector} selector
+ * @param {(selector: string) => boolean} isIgnored
+ * @returns {import('postcss-selector-parser').Selector}
  */
-const zeroSpecificity = () => ({ a: 0, b: 0, c: 0 });
+function removeIgnoredNodes(selector, isIgnored) {
+	const clone = selector.clone();
+
+	clone.walk((node) => {
+		let value = '';
+
+		switch (node.type) {
+			case 'attribute':
+			case 'class':
+			case 'id':
+			case 'tag':
+				value = node.toString();
+				break;
+			case 'pseudo':
+				value = node.value;
+				break;
+
+			default:
+				return
+		}
+
+		if (!value) {
+			return;
+		}
+
+		if (isIgnored(value)) {
+			node.replaceWith(selectorParser.universal({
+				value: '*',
+				sourceIndex: node.sourceIndex,
+				source: node.source,
+			}));
+		}
+	});
+
+	return clone;
+}
 
 /**
- * Calculate the sum of given specificities.
- *
- * @param {Specificity[]} specificities
- * @returns {Specificity}
+ * @typedef {import('postcss-selector-parser').Selector} Selector
+ * @typedef {import('postcss-selector-parser').Root} SelectorRoot
+ * @param {import('postcss').Rule} rule
+ * @param {import('stylelint').PostcssResult} result
+ * @param {(selector: string) => boolean} isIgnored
+ * @returns {Array<{selector: Selector, resolvedSelectors: Array<Selector>}> | undefined}
  */
-const specificitySum = (specificities) => {
-	const sum = zeroSpecificity();
+function gatherSelectorsForRule(rule, result, isIgnored) {
```

Comment:
This should be a shared helper.

It walks the `postcss` AST to gather all the ancestor nodes and resolves all the nested selectors.